third_party/amd/lib/TritonAMDGPUToLLVM/OptimizeLDSUsage.cpp

```diff
@@ -0,0 +1,248 @@
+      tmpLayouts.push_back(fallbackLayout);
+    }
+
+    unsigned minLDSUsage = 2 * LDSSize;
+    int minIdx = -1;
+    for (int i = 0; i < tmpLayouts.size(); i++) {
+      auto resources = mlir::triton::AMD::estimateResourcesForReplacement(
+          builder, cvtOp, tmpLayouts[i]);
+      // TODO analyze performance along with LDS consumption
+      if (resources.LDS < minLDSUsage) {
+        minLDSUsage = resources.LDS;
+        minIdx = i;
+      }
+    }
+
+    if (minIdx == -1 || minLDSUsage > targetLDSSize) {
+      return;
+    }
+
+    triton::gpu::ConvertLayoutOp tmpCvt;
+    triton::gpu::ConvertLayoutOp newEpilogueCvt;
+
+    assert(minIdx >= 0 && minIdx < tmpLayouts.size());
+    auto tmpLayout = tmpLayouts[minIdx];
+    std::tie(tmpCvt, newEpilogueCvt) =
+        mlir::triton::AMD::createNewConvertOps(builder, cvtOp, tmpLayout);
+
+    cvtOp.replaceAllUsesWith(newEpilogueCvt.getResult());
+    cvtOp.erase();
+  }
+
+  struct LDSBottleneckOperation {
+    triton::gpu::ConvertLayoutOp op;
+    int64_t LDSSizeTarget;
+  };
+
+  /**
+   * Assuming that all buffer above scratch buffer in memory space can be
+   * shifted down in memory, this function gives optimistic estimation of memory
+   * space available for scratch buffer.
+   */
+  int64_t
+  computeMaxScratchBufferSize(triton::gpu::ConvertLayoutOp op,
+                              Allocation *allocation,
+                              ArrayRef<Allocation::BufferId> liveBuffers) {
+    int totalSize = 0;
+    auto scratchBufferId = allocation->getBufferId(op.getOperation());
+    int64_t scratchBufferSize = allocation->getAllocatedSize(scratchBufferId);
+    size_t totalLDSConsumption = 0;
+    for (auto buf : liveBuffers)
+      totalLDSConsumption = std::max(
+          totalLDSConsumption, allocation->getAllocatedInterval(buf).end());
+    int64_t freeRequired = totalLDSConsumption - LDSSize;
+    auto maxScratchSize = std::max(0l, scratchBufferSize - freeRequired);
+    return maxScratchSize;
+  }
+
+  SmallVector<LDSBottleneckOperation>
+  findLDSBottleneckLayoutConvert(ModuleAllocation &allocAnalysis,
+                                 FunctionOpInterface func) {
+    SmallVector<LDSBottleneckOperation> candidates;
+    auto funcAnalysis = allocAnalysis.getFuncData(func);
+    auto liveBuffers = funcAnalysis->getLiveBuffers();
+
+    func.walk([&](triton::gpu::ConvertLayoutOp cvtOp) -> void {
+      auto opBuffer = funcAnalysis->getBufferId(cvtOp.getOperation());
+      assert(opBuffer != Allocation::InvalidBufferId);
+      for (auto bufId : liveBuffers[cvtOp]) {
+        auto offset = funcAnalysis->getOffset(bufId);
+        auto size = funcAnalysis->getAllocatedSize(bufId);
+        if (offset + size > LDSSize) {
+          auto maxScratchBufferSize = computeMaxScratchBufferSize(
+              cvtOp, funcAnalysis, liveBuffers[cvtOp]);
+          candidates.push_back({cvtOp, maxScratchBufferSize});
```

Comment:
This function is very confusing to me.
1. Why do we need `opBuffer`? Just to check it's valid?
2. Does `liveBuffers[cvtOp]` include `opBuffer`? To put it another way, does one of the `bufId`'s for the scratch buffer allocated for this cvtOp?
3. It seems to me that this function assumes that there is at most one extra buffer that can overlap with the buffer for this cvtOp? If there are more live buffers that overlap with this cvtOp, we should still only push `cvtOp` into `candidates` once, but compute `maxScratchBufferSize` based on all overlapped live buffers.

Reply:
> Why do we need opBuffer? Just to check it's valid?

Sorry, this is reminder after refactoring, I used to pass it to `computeMaxScratchBufferSize`, but then start compute it inside function.

> Does liveBuffers[cvtOp] include opBuffer? To put it another way, does one of the bufId's for the scratch buffer allocated for this cvtOp?

Yes, scratch buffer is the same as "long-living" buffers, the only difference, that it's live time is limited to one operation.

> It seems to me that this function assumes that there is at most one extra buffer that can overlap with the buffer for this cvtOp? If there are more live buffers that overlap with this cvtOp, we should still only push cvtOp into candidates once, but compute maxScratchBufferSize based on all overlapped live buffers.

No, there could be any number of buffers with live-time overlapping with scratch buffer.

let me remove loop from this function, it should make algorithm clearer.